lib/Dialect/FIRRTL/Transforms/Dedup.cpp

```diff
@@ -856,6 +857,35 @@ void fixupAllModules(InstanceGraph &instanceGraph) {
   }
 }
 
+/// A DenseMapInfo implementation for llvm::SHA256 hashes, which are represented
+/// as std::array<uint8_t, 32>. This allows us to create DenseMaps with SHA256
+/// hashes as keys.
+struct SHA256HashDenseMapInfo {
+  static inline std::array<uint8_t, 32> getEmptyKey() {
+    std::array<uint8_t, 32> key;
+    std::fill(key.begin(), key.end(), ~0);
+    return key;
+  }
+  static inline std::array<uint8_t, 32> getTombstoneKey() {
+    std::array<uint8_t, 32> key;
+    std::fill(key.begin(), key.end(), ~0 - 1);
+    return key;
+  }
+
+  static unsigned getHashValue(const std::array<uint8_t, 32> &val) {
+    // We assume SHA256 is already a good hash and just truncate down to the
+    // number of bytes we need for DenseMap.
+    unsigned hash;
+    std::memcpy(&hash, &val, sizeof(unsigned));
```

Comment:
&val -> val.data()

Reply:
Thanks, fixed in an amended commit. Having made this mistake two PRs in a row, I have now learned to always use `.data()` and stop assuming that `std::array` data can be safely gotten from its pointer address.